src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -221,13 +233,39 @@ private static bool TryParseStatusLine(HttpResponseMessage response, string resp
             // Parse first 3 characters after a space as status code
             // TODO: Parsing errors are treated as fatal. Find right behaviour
             int statusCodeLength = 3;
-            int codeIndex = responseHeader.IndexOf(' ') + 1;
-            string strStatusCode = responseHeader.Substring(codeIndex, statusCodeLength);
-            response.StatusCode = (HttpStatusCode)(int.Parse(strStatusCode));
 
-            // The remaining string after the space is the reason phrase
-            response.ReasonPhrase = responseHeader.Substring(codeIndex + statusCodeLength + 1);
+            int spaceIndex = responseHeader.IndexOf(spaceChar);
+
+            if (spaceIndex > -1)
+            {
+                int codeStartIndex = spaceIndex + 1;
+                int codeEndIndex = codeStartIndex + statusCodeLength;
+
+                if (codeEndIndex <= responseHeaderLength)
+                {
+                    string strStatusCode = responseHeader.Substring(codeStartIndex, statusCodeLength);
+                    response.StatusCode = (HttpStatusCode)(int.Parse(strStatusCode));
+
+                    int reasonPhraseIndex = codeEndIndex + 1;
 
+                    if (reasonPhraseIndex < responseHeaderLength && char.Equals(responseHeader[codeEndIndex], spaceChar))
+                    {
+                        string reasonPhrase = responseHeader.Substring(reasonPhraseIndex);
+
+                        //if reasonPhrase contains any of new line character, then consider reasonPhrase till new line.
+                        int newLineCharIndex = reasonPhrase.IndexOfAny(s_newLineCharArray);
+
+                        if (newLineCharIndex > -1)
+                        {
+                            reasonPhrase = reasonPhrase.Substring(0, newLineCharIndex);
+                        }
+
+                        response.ReasonPhrase = reasonPhrase;
```

Comment:
It'd be better to do just a single Substring rather than two, such that there's only ever at most one string allocation for either case, e.g.
```C#
int newLineCharIndex = responseHeader.IndexOfAny(s_newLineCharArray, reasonPhraseIndex);
int reasonPhraseEnd = newLineCharIndex >= 0 ? newLineCharIndex : responseHeader.Length;
response.ReasonPhrase = responseHeader.Substring(reasonPhraseIndex, reasonPhraseEnd - reasonPhraseIndex);
```